mininterface/textual_interface/widgets.py

```diff
@@ -33,8 +33,91 @@ def get_ui_value(self):
+            # Case 1: Called with a Tag object
+            tag = value_or_tag
+            self._link = tag
+            initial_value = ""
+            if hasattr(tag, 'val') and tag.val is not None:
+                if isinstance(tag.val, list):
+                    initial_value = ", ".join(str(p) for p in tag.val)
+                else:
+                    initial_value = str(tag.val)
+            self.value = initial_value
+        else:
+            # Case 2: Called with a direct value (str, int, etc.)
+            self.value = str(value_or_tag)
+            # The _link will be set later by the calling function (widgetize)
+
+    def _convert_value(self, value: str):
+        """Convert the string value to the appropriate type."""
+        if not value:
+            return None
+
+        value = value.strip()
+        if not value:
+            return None
+
+        # If _link isn't set yet, return the value as is
+        if not hasattr(self, '_link'):
+            return value
+
+        # Special handling for PathTag
+        if isinstance(self._link, PathTag):
+            try:
+                if getattr(self._link, 'multiple', False):
+                    paths = [p.strip() for p in value.split(',') if p.strip()]
+                    if not paths:
+                        return None
+                    # Make sure all paths exist with Path objects
+                    return [Path(p) for p in paths]
+                # Single path case
+                return Path(value)
+            except Exception:  # noqa
+                # Always return the original string value for PathTag if conversion fails
+                # This prevents "Type must be str" error
+                if getattr(self._link, 'multiple', False):
+                    return [value.strip()] if value.strip() else None
+                return value
+
+        try:
+            # Get the expected type from the tag's annotation
+            expected_type = getattr(self._link, 'annotation', str)
+
+            if expected_type == int:
+                return int(value)
+            elif expected_type == float:
+                return float(value)
+            else:
+                return value
+        except (ValueError, TypeError):
+            return None
+
+    def on_blur(self, event: events.Blur) -> None:
+        # Only trigger if we can convert the value successfully
+        if self._convert_value(self.value) is not None:
+            self.trigger_change()
+
+    def on_input_changed(self, event: Input.Changed) -> None:
+        # Only trigger if we can convert the value successfully
+        if self._convert_value(self.value) is not None:
+            self.trigger_change()
+
+    def on_input_submitted(self, event: Input.Submitted) -> None:
+        # This is triggered when Enter is pressed
+        if self._convert_value(self.value) is not None:
+            self.trigger_change()
+            if hasattr(self._link, 'facet'):
```

Comment:
Facet is an optional attribute, so it is always there but might be None.

Hence `if self._link.facet`.